core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -203,17 +204,19 @@ public static State fromString(String state) {
     private final HashFunction routingHashFunction;
     private final boolean useTypeForRouting;
 
-    private IndexMetaData(String index, long version, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings, ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs) {
+    private IndexMetaData(String index, long version, int[] primaryTerms, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings, ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs) {
         if (settings.getAsInt(SETTING_NUMBER_OF_SHARDS, null) == null) {
             throw new IllegalArgumentException("must specify numberOfShards for index [" + index + "]");
         }
         if (settings.getAsInt(SETTING_NUMBER_OF_REPLICAS, null) == null) {
             throw new IllegalArgumentException("must specify numberOfReplicas for index [" + index + "]");
         }
+        this.settings = settings;
         this.index = index;
         this.version = version;
+        this.primaryTerms = primaryTerms;
+        assert primaryTerms.length == numberOfShards();
```

Comment:
> Are we really okay with all of this repeated parsing and boxing and unboxing

I'm +1 on changing it - but it's a different change. This is not how performant code though  :)

> Also, I think it's best if in general we not invoke instance methods in constructors 

+1 here too - but this is just an assertion - I felt it was worth it.